### PR TITLE
New grammar construction modes for SyGuS

### PR DESCRIPTION
**Pull request by ajreynol**
This introduces an option `--sygus-grammar-cons` to specify a mode for SyGuS grammar construction. These mode correspond to whether any constant constructors are included, or whether any constants are used to specify coefficients on monomials.

### CODE REVIEW

File: src/theory/quantifiers/sygus/sygus_grammar_cons.cpp [hunk 1]
```diff
@@ -822,8 +874,221 @@ void CegGrammarConstructor::mkSygusDefaultGrammar(
...
+          // this is already corresponds to the correct sygus datatype type
+          TypeNode atype = sdc.d_argTypes[j];
+          opCArgs.push_back(atype);
+          // get the builtin type
+          TypeNode btype = sygus_to_builtin[atype];
+          opLArgs.push_back(nm->mkBoundVar(btype));
+        }
+        // Do beta reduction on the operator so that its arguments match the
+        // fresh variables of the lambda we are constructing.
+        sop = datatypes::utils::mkSygusTerm(sop, opLArgs);
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/sygus_grammar_cons.cpp]: Do you mean that `sop` is a lambda? Isn't it coming directly from the datatype? Overall this part is hard to follow

File: src/theory/quantifiers/sygus/sygus_grammar_cons.cpp [hunk 2]
```diff
@@ -822,8 +874,221 @@ void CegGrammarConstructor::mkSygusDefaultGrammar(
...
+      spc = std::make_shared<printer::SygusExprPrintCallback>(ops.toExpr(),
+                                                              lambdaVarsExpr);
+      Trace("sygus-grammar-def") << "any term operator is " << op << std::endl;
+      // make the any term datatype, add to back
+      // do not consider the exclusion criteria of the generator
+      sdts[iat].d_sdt.addConstructor(op, "polynomial", cargsAnyTerm, spc);
+    }
+    else
+    {
+      // add the any constant constructor as a separator constructor
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/sygus_grammar_cons.cpp]: separated?

File: src/options/options_handler.cpp [hunk 3]
```diff
@@ -569,6 +569,27 @@ auto (default) \n\
 \n\
 ";
 
+const std::string OptionsHandler::s_sygusGrammarConsHelp =
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/options/options_handler.cpp]: There are already the possibilities of having more than one option that affects the grammar (see --sygus-add-const-grammar, --sygus-templ-embed-grammar, etc.). The purpose of this option is to set a baseline that those other options modify.

File: src/options/options_handler.cpp [hunk 4]
```diff
@@ -569,6 +569,27 @@ auto (default) \n\
...
++ Use symoblic constant constructors.\n\
+\n\
+any-term  \n\
++ When applicable, use constructors corresponding to any symbolic term.\n\
+This option enables a sum-of-monomials grammar for arithmetic.\n\
+\n\
+any-term-concise  \n\
++ When applicable, use constructors corresponding to any symbolic term,\n\
+favoring conciseness over generality. This option enables a polynomial\n\
+grammar for arithmetic.\n\
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/options/options_handler.cpp]: No, then they would need to be separate options.  The point here is to distinguish what level of symbolic constants to include. The possibilities here are fixed to be mutually exclusive.

File: src/theory/quantifiers/sygus/sygus_grammar_cons.cpp [hunk 5]
```diff
@@ -812,7 +864,7 @@ void CegGrammarConstructor::mkSygusDefaultGrammar(
           << types[i] << std::endl;
     }
   }
-  // make datatypes
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/sygus/sygus_grammar_cons.cpp]: Make is actually inaccurate, it should say "initialize datatypes", I've updated.

File: src/options/options_handler.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (HanielB) [reply src/options/options_handler.cpp]: What is not clear however is that some of these may be activated by the others. You could add comments to say that "in this situation, this reduces to that".
